zerver/tests/test_typing.py

```diff
@@ -136,6 +203,40 @@ def test_send_notification_to_self_event(self) -> None:
         self.assertEqual(event['type'], 'typing')
         self.assertEqual(event['op'], 'start')
 
+    def test_send_notification_to_self_by_user_id_event(self) -> None:
+        """
+        Sending typing notification to yourself (using user IDs)
+        is successful.
+        """
+        user = self.example_user('hamlet')
+        email = user.email
+        expected_recipient_emails = set([email])
+        expected_recipient_ids = set([user.id])
+        events = []  # type: List[Mapping[str, Any]]
+        with tornado_redirected_to_list(events):
+            result = self.api_post(
+                email,
+                '/api/v1/typing',
+                {
+                    'to': ujson.dumps([user.id]),
+                    'op': 'start'
+                }
+            )
+        self.assert_json_success(result)
+        self.assertEqual(len(events), 1)
+
+        event = events[0]['event']
+        event_recipient_emails = set(user['email'] for user in event['recipients'])
+        event_user_ids = set(events[0]['users'])
+        event_recipient_user_ids = set(user['user_id'] for user in event['recipients'])
+
+        self.assertEqual(expected_recipient_ids, event_recipient_user_ids)
+        self.assertEqual(expected_recipient_ids, event_user_ids)
+        self.assertEqual(event_recipient_emails, expected_recipient_emails)
+        self.assertEqual(event['sender']['email'], email)
+        self.assertEqual(event['type'], 'typing')
+        self.assertEqual(event['op'], 'start')
```

Comment:
I'm going to merge this, but I think given how repetitive these tests are, it's probably worth extracting a helper function for the common code as a follow-up.